contracts/EscrowAndStaking/CrossChainCollector.sol

```diff
@@ -0,0 +1,408 @@
+            } else if (masterCollector == sender) {
+                _updateRewardsOnStakingContract(any2EvmMessage.data);
+            }
+        }
+        emit MessageReceived(
+            any2EvmMessage.messageId,
+            any2EvmMessage.sourceChainSelector, // fetch the source chain identifier (aka selector)
+            abi.decode(any2EvmMessage.sender, (address)), // abi-decoding of the sender address,
+            any2EvmMessage.data
+        );
+    }
+
+    function _calculateRewards(bytes memory data, uint _chainSelector) internal {
+        (uint stakedAmount, uint escrowedAmount, uint bonusPoints, uint revShare) = abi.decode(
+            data,
+            (uint, uint, uint, uint)
+        );
+        _storeRewards(_chainSelector, stakedAmount, escrowedAmount, bonusPoints, revShare);
+    }
+
+    function _storeRewards(
+        uint _chainSelector,
+        uint _stakedAmount,
+        uint _escrowedAmount,
+        uint _bonusPoints,
+        uint _revShare
+    ) internal {
+        chainStakedAmountInPeriod[period][_chainSelector] = _stakedAmount;
+        chainEscrowedAmountInPeriod[period][_chainSelector] = _escrowedAmount;
+        chainBonusPointsInPeriod[period][_chainSelector] = _bonusPoints;
+        chainRevenueInPeriod[period][_chainSelector] = _revShare;
+
+        calculatedStakedAmountForPeriod[period] += _stakedAmount;
+        calculatedEscrowedAmountForPeriod[period] += _escrowedAmount;
+        calculatedBonusPointsForPeriod[period] += _bonusPoints;
+        calculatedRevenueForPeriod[period] += _revShare;
+    }
+
+    function _broadcastMessageToAll() internal {
+        uint chainBaseRewards;
+        uint chainExtraRewards;
+        uint revShare;
+        bytes memory message;
+
+        for (uint i = 0; i < numOfActiveCollectors; i++) {
+            chainBaseRewards =
+                ((chainStakedAmountInPeriod[period][chainSelector[i]] +
+                    chainEscrowedAmountInPeriod[period][chainSelector[i]]) * baseRewardsPerPeriod) /
+                (calculatedStakedAmountForPeriod[period] + calculatedEscrowedAmountForPeriod[period]);
+
+            chainExtraRewards =
+                (chainBonusPointsInPeriod[period][chainSelector[i]] * extraRewardsPerPeriod) /
+                (calculatedBonusPointsForPeriod[period]);
+
+            revShare =
+                ((chainStakedAmountInPeriod[period][chainSelector[i]] +
+                    chainEscrowedAmountInPeriod[period][chainSelector[i]]) * calculatedRevenueForPeriod[period]) /
+                (calculatedStakedAmountForPeriod[period] + calculatedEscrowedAmountForPeriod[period]);
+
+            chainBaseRewardsInPeriod[period][chainSelector[i]] = chainBaseRewards;
+            chainExtraRewardsInPeriod[period][chainSelector[i]] = chainExtraRewards;
+            chainRevenueShareInPeriod[period][chainSelector[i]] = revShare;
+
+            message = abi.encode(chainBaseRewards, chainExtraRewards, revShare);
+            if (i == 0) {
```

Comment:
how is this going to be handled if we need to change the master node?

Reply:
changed to check if it is the masterCollector selector